ipfsconn/ipfshttp/ipfshttp.go

```diff
@@ -450,6 +483,79 @@ func checkResponse(path string, code int, body []byte) error {
 	return fmt.Errorf("IPFS-post '%s' unsuccessful: %d: %s", path, code, body)
 }
 
+func handleRefsProgress(dec *json.Decoder, reset chan int) error {
+	var last string
+	for {
+		var ref ipfsRefsResp
+		if err := dec.Decode(&ref); err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		//logger.Infof("Ref %s", ref.Ref)
+		if ref.Err != "" {
+			logger.Error(ref.Err)
+		}
+
+		if ref.Ref != last {
+			last = ref.Ref
+			reset <- 1
+		}
+	}
+
+	return nil
+}
+
+func handlePinsProgress(dec *json.Decoder, reset chan int) error {
+	var progress int
+	for {
+		var pins ipfsPinsResp
+		if err := dec.Decode(&pins); err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+		//logger.Infof("Progress: %d", pins.Progress)
+		if pins.Progress > progress {
+			progress = pins.Progress
+			reset <- 1
+		}
+	}
+
+	return nil
+}
+
+func checkTimeout(ctx context.Context, cancel context.CancelFunc, timer *time.Timer, timeout time.Duration, reset chan int) {
+	var done bool
+	go func() {
+		for {
+			select {
+			case <-reset:
+				{
+					if timer.Stop() {
+						timer.Reset(timeout)
```

Comment:
We could end up being ressetting a timer every few milliseconds in default the case of `refs`, or every second or so with `pin add`. Would it be better to have a ticker that drains the reset channel once a second (or even once every 5) and then resets the timer? The timeouts here are at least minutes long. The sending part must not block on a full channel.

Reply:
> Would it be better to have a ticker that drains the reset channel once a second (or even once every 5)

This would add blocking time of one or five seconds because of the channels. I have some ideas, let me try.

Reply:
You should send like:

```
select {
case reset <- struct{}{}:
default:
}
```

so it doesn't block

Reply:
I have tried https://github.com/ipfs/ipfs-cluster/pull/738/commits/d640fa8157d775bc3d14d541777e0fd4cc60f516
This reduces the number of times that we reset, but it still looks really horrible. There should be a better way.